### PR TITLE
Added check status logic when updating acq and related entities | psp-7006

### PR DESCRIPTION
**Pull request by FuriousLlama**

### CODE REVIEW

**Comment by devinleighsmith** [comment]
@FuriousLlama I checked this out, but I don't see this anywhere:

Edit Button Is Hidden

 GIVEN
User does not have permissions to edit a tab
WHEN
User views restricted tab
THEN
System will show a tooltip that indicates why editing is not available.
"The file you are viewing is in a non-editable state. Change the file status to active or draft to allow editing."

**Comment by devinleighsmith** [comment]
Hmm, the story does not go into how to control the behaviour of making an agreement editable/non-editable when it is swapped from draft to final in a file that is already in the final status. I do think it is a bit jarring to disable the fields the instant the status change happens (before the agreement is saved). Could that be changed to only make non-editable when the agreement status is final AND there is no id (indicating an unsaved agreement)?

File: source/backend/api/Services/AcquisitionFileService.cs [hunk 1]
```diff
@@ -240,6 +244,12 @@ public PimsAcquisitionFile Update(PimsAcquisitionFile acquisitionFile, IEnumerab
             ValidateVersion(acquisitionFile.Internal_Id, acquisitionFile.ConcurrencyControlNumber);
             ValidateDrafts(acquisitionFile.Internal_Id);
 
+            AcqusitionStatusTypes? currentAcquisitionStatus = GetCurrentAcquisitionStatus(acquisitionFile.Internal_Id);
```
Comment (devinleighsmith) [inline source/backend/api/Services/AcquisitionFileService.cs:247]: nit: acqusition -> acquisition

File: source/backend/api/Constants/AcqusitionStatusTypes.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
+using System.Runtime.Serialization;
+using System.Text.Json.Serialization;
+
+namespace Pims.Api.Constants
+{
+    [JsonConverter(typeof(JsonStringEnumMemberConverter))]
+    public enum AcqusitionStatusTypes
```
Comment (devinleighsmith) [inline source/backend/api/Constants/AcqusitionStatusTypes.cs:7]: nit: acqusition

File: source/backend/api/Services/AcquisitionFileService.cs [hunk 3]
```diff
@@ -393,6 +409,31 @@ public IEnumerable<PimsAgreement> UpdateAgreements(long acquisitionFileId, List<
...
             _user.ThrowInvalidAccessToAcquisitionFile(_userRepository, _acqFileRepository, acquisitionFileId);
 
+            var currentAcquisitionStatus = GetCurrentAcquisitionStatus(acquisitionFileId);
+
+            var currentAgreements = _agreementRepository.GetAgreementsByAcquisitionFile(acquisitionFileId);
+
+            var toBeUpdated = currentAgreements.Where(ca => agreements.Any(na => ca.AgreementId == na.AgreementId && !ca.IsEqual(na)));
+            var toBeDeleted = currentAgreements.Where(ca => !agreements.Any(na => ca.AgreementId == na.AgreementId));
+
+            foreach (var agreement in toBeUpdated)
```
Comment (devinleighsmith) [inline source/backend/api/Services/AcquisitionFileService.cs:419]: is there a reason that agreements being edited and agreements being deleted are being split out into their own loops? Seems like the business logic that is being executed is identical, sort of adds an implication that updates should not be handles the same as deletes.

File: source/backend/api/Services/AcquisitionFileService.cs
Comment (devinleighsmith) [inline source/backend/api/Services/AcquisitionFileService.cs:1]: It looks like you added an explicit check/throw for stakeholders and checklists even though those errors will not get thrown in current state. I'm fine with that as it provides a consistent pattern for update logic, but doesn't that mean you should add a check for updateProperties as well?

File: source/backend/api/Services/CompensationRequisitionService.cs [hunk 4]
```diff
@@ -59,8 +62,15 @@ public PimsCompensationRequisition Update(PimsCompensationRequisition compensati
 
             var currentCompensation = _compensationRequisitionRepository.GetById(compensationRequisition.CompensationRequisitionId);
 
-            CheckDraftStatusUpdateAuthorized(currentCompensation.IsDraft, compensationRequisition.IsDraft);
-            CheckTotalAllowableCompensation(compensationRequisition.AcquisitionFileId, compensationRequisition);
+            var currentAcquisitionFile = _acqFileRepository.GetById(currentCompensation.AcquisitionFileId);
+            var currentAcquisitionStatus = Enum.Parse<AcqusitionStatusTypes>(currentAcquisitionFile.AcquisitionFileStatusTypeCode);
+
+            if (!_statusSolver.CanEditOrDeleteCompensation(currentAcquisitionStatus, currentCompensation.IsDraft) && !_user.HasPermission(Permissions.SystemAdmin))
```
Comment (devinleighsmith) [inline source/backend/api/Services/CompensationRequisitionService.cs:68]: I actually don't see a requirement for an admin to be able to override this in the story, can you point that out to me?

File: source/backend/api/Services/CompensationRequisitionService.cs [hunk 5]
```diff
@@ -90,6 +100,15 @@ public bool DeleteCompensation(long compensationId)
             _logger.LogInformation("Deleting compensation with id ...", compensationId);
             _user.ThrowIfNotAuthorized(Permissions.CompensationRequisitionDelete, Permissions.AcquisitionFileEdit);
 
+            var currentCompensation = _compensationRequisitionRepository.GetById(compensationId);
+
+            var currentAcqusitionStatus = GetCurrentAcqusitionStatus(currentCompensation.AcquisitionFileId);
```
Comment (devinleighsmith) [inline source/backend/api/Services/CompensationRequisitionService.cs:105]: nit: getcurrent acqusition

File: source/backend/api/Services/AcquisitionFileService.cs
Comment (FuriousLlama) [reply source/backend/api/Services/AcquisitionFileService.cs:1]: From what I gather on the confluence page the properties are allowed to be updated at all times (all statues)

File: source/backend/api/Services/CompensationRequisitionService.cs (hunk 4 above)
Comment (FuriousLlama) [reply source/backend/api/Services/CompensationRequisitionService.cs:68]: It's from a different story. I figured those where additive instead of replaced.

File: source/backend/api/Services/AcquisitionFileService.cs
Comment (devinleighsmith) [reply source/backend/api/Services/AcquisitionFileService.cs:1]: right, but the same applies to stakeholders and checklists, why have explicit logic for those and not for properties?

File: source/backend/entities/Partials/Agreement.cs [hunk 6]
```diff
@@ -11,5 +11,28 @@ public partial class PimsAgreement : StandardIdentityBaseAppEntity<long>, IBaseA
         [NotMapped]
         public override long Internal_Id { get => this.AgreementId; set => this.AgreementId = value; }
         #endregion
+
+        public bool IsEqual(PimsAgreement other)
```
Comment (devinleighsmith) [inline source/backend/entities/Partials/Agreement.cs:15]: In my experience isEqual methods generally handle being passed a null other value.

File: source/backend/tests/core/TestHelper.cs [hunk 7]
```diff
@@ -11,8 +11,10 @@
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Options;
 using Moq;
+using Pims.Api.Services;
 using Pims.Dal;
 using Pims.Dal.Configuration.Generators;
+using Pims.Dal.Entities;
```
Comment (devinleighsmith) [inline source/backend/tests/core/TestHelper.cs:17]: this seems odd, looks like you are adding imports but then not using them anywhere?

File: source/frontend/src/features/mapSideBar/acquisition/tabs/checklist/detail/AcquisitionChecklistView.tsx [hunk 8]
```diff
@@ -38,10 +39,12 @@ export const AcquisitionChecklistView: React.FC<IAcquisitionChecklistViewProps>
   const checklist = acquisitionFile?.acquisitionFileChecklist || [];
   const lastUpdated = lastModifiedBy(checklist);
 
+  const statusSolver = new StatusUpdateSolver(acquisitionFile);
+
   return (
     <StyledSummarySection>
       <StyledEditWrapper className="mr-3 my-1">
-        {keycloak.hasClaim(Claims.ACQUISITION_EDIT) && acquisitionFile !== undefined ? (
```
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/checklist/detail/AcquisitionChecklistView.tsx:44]: are we ok removing that undefined acquisitionFile check?

File: source/frontend/src/features/mapSideBar/acquisition/tabs/compensation/detail/CompensationRequisitionDetailContainer.tsx [hunk 9]
```diff
@@ -91,11 +91,10 @@ export const CompensationRequisitionDetailContainer: React.FunctionComponent<
   return compensation ? (
     <View
       loading={loading || loadingPerson || loadingOrganization}
+      acquisitionFile={acquisitionFile}
       compensation={compensation}
       compensationContactPerson={payeePerson}
       compensationContactOrganization={payeeOrganization}
-      acqFileProject={acquisitionFile?.project}
```
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/compensation/detail/CompensationRequisitionDetailContainer.tsx:97]: huh, I wonder why these were split out like this?

File: source/frontend/src/features/mapSideBar/acquisition/tabs/stakeholders/detail/StakeHolderContainer.test.tsx [hunk 10]
```diff
@@ -83,101 +78,4 @@ describe('StakeHolderContainer component', () => {
     const { asFragment } = setup({});
     expect(asFragment()).toMatchSnapshot();
   });
-
-  it('groups multiple interest holder properties by acquisition file id', async () => {
```
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/stakeholders/detail/StakeHolderContainer.test.tsx:87]: ?

File: source/frontend/src/features/mapSideBar/acquisition/tabs/stakeholders/detail/StakeHolderView.test.tsx [hunk 11]
```diff
@@ -128,11 +150,13 @@ describe('StakeHolderView component', () => {
...
     );
 
+    organizerMock.getInterestProperties.mockReturnValue([model]);
+
     model.identifier = 'PID: 025-196-375';
     const { getByText } = setup({
       props: {
-        groupedInterestProperties: [],
-        groupedNonInterestProperties: [model],
+        //groupedInterestProperties: [],
```
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/stakeholders/detail/StakeHolderView.test.tsx:158]: ? if this is a wip, please fix. if not, can you remove these comments?

File: source/frontend/src/features/mapSideBar/acquisition/tabs/stakeholders/detail/StakeHolderContainer.tsx [hunk 12]
```diff
@@ -34,75 +31,14 @@ export const StakeHolderContainer: React.FunctionComponent<IStakeHolderContainer
...
-      )}
-      legacyStakeHolders={acquisitionFile.legacyStakeholders ?? []}
+      acquisitionFile={acquisitionFile}
+      interestHolders={apiInterestHolders}
       onEdit={onEdit}
     />
   );
 };
 
-const getGroupedInterestProperties = (
```
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/stakeholders/detail/StakeHolderContainer.tsx:71]: Is this related to this PR? If not, I'd really prefer we separate these kind of changes into their own PR.

File: source/frontend/src/features/mapSideBar/acquisition/tabs/stakeholders/detail/stakeholderOrganizer.ts [hunk 13]
```diff
@@ -0,0 +1,84 @@
+import { Api_AcquisitionFile } from '@/models/api/AcquisitionFile';
```
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/stakeholders/detail/stakeholderOrganizer.ts:1]: This looks a bit like unrelated tech-debt. I prefer this, but feel like this should have been done in a separate PR at least.

**Comment by devinleighsmith** [comment]
I'm seeing a lot of line coverage warnings for these changes. Can you confirm that this PR conforms to our minimum test coverage requirements:
1. one test per business logic.
2. each layer should be mocked, validate that mocked layers are called in the expected manner.

File: source/frontend/src/features/mapSideBar/acquisition/tabs/stakeholders/detail/StakeHolderView.test.tsx (hunk 11 above)
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/stakeholders/detail/StakeHolderView.test.tsx:158]: ups. forgot to cleanup this tests

File: source/frontend/src/features/mapSideBar/acquisition/tabs/checklist/detail/AcquisitionChecklistView.tsx (hunk 8 above)
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/checklist/detail/AcquisitionChecklistView.tsx:44]: the status solver does take into account the acq file being null/undefined

File: source/frontend/src/features/mapSideBar/acquisition/tabs/compensation/detail/CompensationRequisitionDetailContainer.tsx (hunk 9 above)
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/compensation/detail/CompensationRequisitionDetailContainer.tsx:97]: note sure, but now is more cohesive

File: source/frontend/src/features/mapSideBar/acquisition/tabs/stakeholders/detail/StakeHolderContainer.test.tsx (hunk 10 above)
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/stakeholders/detail/StakeHolderContainer.test.tsx:87]: All the tests for the grouping logic was movied into its own test file that only deals with that.

File: source/frontend/src/features/mapSideBar/acquisition/tabs/stakeholders/detail/StakeHolderContainer.tsx (hunk 12 above)
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/stakeholders/detail/StakeHolderContainer.tsx:71]: Thats fair. I rather keep it as is now and in the future create new PRs

File: source/frontend/src/features/mapSideBar/acquisition/tabs/stakeholders/detail/stakeholderOrganizer.ts (hunk 13 above)
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/stakeholders/detail/stakeholderOrganizer.ts:1]: True, I refactored to leverage similar props being passed. It also contains the grouping logic instead of being part of the view

**Comment by FuriousLlama** [comment]
> I'm seeing a lot of line coverage warnings for these changes. Can you confirm that this PR conforms to our minimum test coverage requirements:
> 
> 1. one test per business logic.
> 2. each layer should be mocked, validate that mocked layers are called in the expected manner.

The backend logic has very thorough tests. I can add some to the front end too for the solver, but otherwise, it should have the same amount of tests.

File: source/backend/entities/Partials/Agreement.cs (hunk 6 above)
Comment (FuriousLlama) [reply source/backend/entities/Partials/Agreement.cs:15]: That's a good point, however, this one is not a Comparable but rather a helper method. I can make it more standard too but I am not sure we gain much from it.

File: source/backend/api/Services/AcquisitionFileService.cs (hunk 3 above)
Comment (FuriousLlama) [reply source/backend/api/Services/AcquisitionFileService.cs:419]: Although true, doing the loop on each would be the same as doing the loop on an aggregated collection.

File: source/backend/api/Services/AcquisitionFileService.cs
Comment (FuriousLlama) [reply source/backend/api/Services/AcquisitionFileService.cs:1]: Ah, I understand what you are saying now. Initially I had it like that, but it felt like adding those checks was adding unnecessary complexity. If you feel strongly about it I can add them back.

**Review by devinleighsmith — APPROVED** [review]
Approved, but you are going to get a test reject from not displaying the warning tooltip in place of the edit icon.

**Comment by FuriousLlama** [comment]
> Approved, but you are going to get a test reject from not displaying the warning tooltip in place of the edit icon.

Tooltip is there :)
